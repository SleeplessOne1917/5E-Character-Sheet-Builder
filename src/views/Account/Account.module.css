.username-blurb {
    font-size: 1.3rem;
    padding: 1rem;
    align-self: center;
}

.username {
    font-size: 1.5rem;
    font-weight: bold;
    color: var(--wine-red)
}

.reset-text {
    font-family: var(--font-fantasy);
    color: var(--davy);
    align-self: center;
}

.form {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-around;
}