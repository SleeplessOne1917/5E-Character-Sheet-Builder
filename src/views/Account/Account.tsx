import Button, { ButtonType } from '../../components/Button/Button';
import { useAppDispatch, useAppSelector } from '../../hooks/reduxHooks';

import CREATE_NEW_PASSWORD from '../../graphql/mutations/user/createNewPassword';
import { Formik } from 'formik';
import MainContent from '../../components/MainContent/MainContent';
import TextInput from '../../components/TextInput/TextInput';
import { ToastType } from '../../types/toast';
import classes from './Account.module.css';
import newPasswordSchema from '../../yup-schemas/newPasswordSchema';
import { show } from '../../redux/features/toast';
import { useMutation } from 'urql';

const Account = () => {
	const username = useAppSelector(state => state.viewer);
	const [_, createNewPassword] = useMutation(CREATE_NEW_PASSWORD);
	const dispatch = useAppDispatch();

	return (
		<MainContent testId="account">
			<h1>Account</h1>
			<div className={classes['username-blurb']}>
				Your username is <span className={classes.username}>{username}</span>
			</div>
			<h2 className={classes['reset-text']}>Create New Password</h2>
			<Formik
				initialValues={{
					currentPassword: '',
					newPassword: '',
					confirmPassword: ''
				}}
				validationSchema={newPasswordSchema}
				onSubmit={async (values, { resetForm }) => {
					const result = await createNewPassword(values);

					if (result.error) {
						dispatch(
							show({
								closeTimeoutSeconds: 10,
								message: result.error.message,
								type: ToastType.error
							})
						);
					} else {
						dispatch(
							show({
								closeTimeoutSeconds: 10,
								message: result.data.createNewPassword,
								type: ToastType.success
							})
						);
					}

					resetForm();
				}}
			>
				{({
					errors,
					touched,
					values,
					handleBlur,
					handleChange,
					handleSubmit,
					setTouched,
					isSubmitting
				}) => (
					<form onSubmit={handleSubmit} className={classes.form}>
						<TextInput
							id="currentPassword"
							label="Current Password"
							onBlur={handleBlur}
							onChange={event => {
								setTouched({ ...touched, currentPassword: false });
								handleChange(event);
							}}
							value={values.currentPassword}
							error={errors.currentPassword}
							touched={touched.currentPassword}
							type="password"
						/>
						<TextInput
							id="newPassword"
							label="New Password"
							onBlur={handleBlur}
							onChange={event => {
								setTouched({ ...touched, newPassword: false });
								handleChange(event);
							}}
							value={values.newPassword}
							error={errors.newPassword}
							touched={touched.newPassword}
							type="validate-password"
						/>
						<TextInput
							id="confirmPassword"
							label="Confirm Password"
							onBlur={handleBlur}
							onChange={event => {
								setTouched({ ...touched, confirmPassword: false });
								handleChange(event);
							}}
							value={values.confirmPassword}
							error={errors.confirmPassword}
							touched={touched.confirmPassword}
							type="password"
						/>
						<div style={{ marginTop: '1rem' }}>
							<Button
								disabled={isSubmitting}
								type={ButtonType.submit}
								size="medium"
								positive
							>
								Create new password
							</Button>
						</div>
					</form>
				)}
			</Formik>
		</MainContent>
	);
};

export default Account;
