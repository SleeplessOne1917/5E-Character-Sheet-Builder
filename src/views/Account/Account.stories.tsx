import { ComponentMeta, ComponentStory } from '@storybook/react';

import Account from './Account';
import { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { getTestStore } from '../../redux/store';
import loggedInViewerMock from '../../mock/loggedInViewerMock';

const MockStore = ({ children }: PropsWithChildren) => (
	<Provider store={getTestStore({ viewer: loggedInViewerMock })}>
		{children}
	</Provider>
);

export default {
	title: 'Views/Account',
	component: Account,
	args: { loading: false },
	decorators: [
		Story => (
			<MockStore>
				<Story />
			</MockStore>
		)
	]
} as ComponentMeta<typeof Account>;

const Template: ComponentStory<typeof Account> = args => <Account {...args} />;

export const Default = Template.bind({});
