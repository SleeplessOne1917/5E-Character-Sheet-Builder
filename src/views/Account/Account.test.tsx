import * as Router from 'next/navigation';
import * as stories from './Account.stories';

import { render, screen } from '@testing-library/react';

import { composeStories } from '@storybook/react';

const { Default } = composeStories(stories);

const mockRouter = (pathname: string) => {
	const useRouter = jest.spyOn(Router, 'useRouter');

	const router = {
		push: () => Promise.resolve(true),
		prefetch: () => Promise.resolve(),
		pathname,
		route: '',
		query: {},
		asPath: '',
		basePath: '',
		isLocaleDomain: true,
		reload: jest.fn(),
		replace: jest.fn(),
		back: jest.fn(),
		beforePopState: jest.fn(),
		events: { emit: jest.fn(), on: jest.fn(), off: jest.fn() },
		isFallback: false,
		isPreview: false,
		isReady: true,
		forward: () => {}
	};
	useRouter.mockReturnValue(router);
};

it('renders correctly', () => {
	mockRouter('/account');
	render(<Default />);

	expect(screen.getByTestId('account')).toMatchSnapshot();
});
