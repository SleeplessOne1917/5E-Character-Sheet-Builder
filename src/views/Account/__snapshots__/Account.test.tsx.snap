// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<main
  class="main"
  data-testid="account"
>
  <div
    class="content"
  >
    <h1>
      Account
    </h1>
    <div
      class="username-blurb"
    >
      Your username is 
      <span
        class="username"
      />
    </div>
    <h2
      class="reset-text"
    >
      Create New Password
    </h2>
    <form
      class="form"
    >
      <div
        class="input-and-error-container"
        data-testid="text-input"
      >
        <div
          class="input-and-label-container"
        >
          <div
            class="input-container"
          >
            <input
              class="input"
              id="currentPassword"
              name="currentPassword"
              type="password"
              value=""
            />
            <svg
              aria-hidden="false"
              aria-label="Show Current Password"
              class="eye"
              fill="currentColor"
              tabindex="0"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 12.5a2.5 2.5 0 100-5 2.5 2.5 0 000 5z"
              />
              <path
                clip-rule="evenodd"
                d="M.664 10.59a1.651 1.651 0 010-1.186A10.004 10.004 0 0110 3c4.257 0 7.893 2.66 9.336 6.41.147.381.146.804 0 1.186A10.004 10.004 0 0110 17c-4.257 0-7.893-2.66-9.336-6.41zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
          <label
            class="label"
            for="currentPassword"
          >
            Current Password
          </label>
        </div>
      </div>
      <div
        class="input-and-error-container"
        data-testid="text-input"
      >
        <div
          class="input-and-label-container"
        >
          <div
            class="input-container"
          >
            <input
              class="input"
              id="newPassword"
              name="newPassword"
              type="password"
              value=""
            />
            <svg
              aria-hidden="false"
              aria-label="Show New Password"
              class="eye"
              fill="currentColor"
              tabindex="0"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 12.5a2.5 2.5 0 100-5 2.5 2.5 0 000 5z"
              />
              <path
                clip-rule="evenodd"
                d="M.664 10.59a1.651 1.651 0 010-1.186A10.004 10.004 0 0110 3c4.257 0 7.893 2.66 9.336 6.41.147.381.146.804 0 1.186A10.004 10.004 0 0110 17c-4.257 0-7.893-2.66-9.336-6.41zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
          <label
            class="label"
            for="newPassword"
          >
            New Password
          </label>
        </div>
        <div
          class="container"
          data-testid="container"
        >
          <div
            class="validator no-match"
          >
            <svg
              aria-hidden="true"
              class="icon x"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M3.97 3.97a.75.75 0 011.06 0L12 10.94l6.97-6.97a.75.75 0 111.06 1.06L13.06 12l6.97 6.97a.75.75 0 11-1.06 1.06L12 13.06l-6.97 6.97a.75.75 0 01-1.06-1.06L10.94 12 3.97 5.03a.75.75 0 010-1.06z"
                fill-rule="evenodd"
              />
            </svg>
             
            At least 8 characters long
          </div>
          <div
            class="validator no-match"
          >
            <svg
              aria-hidden="true"
              class="icon x"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M3.97 3.97a.75.75 0 011.06 0L12 10.94l6.97-6.97a.75.75 0 111.06 1.06L13.06 12l6.97 6.97a.75.75 0 11-1.06 1.06L12 13.06l-6.97 6.97a.75.75 0 01-1.06-1.06L10.94 12 3.97 5.03a.75.75 0 010-1.06z"
                fill-rule="evenodd"
              />
            </svg>
             
            At least 1 lowercase letter
          </div>
          <div
            class="validator no-match"
          >
            <svg
              aria-hidden="true"
              class="icon x"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M3.97 3.97a.75.75 0 011.06 0L12 10.94l6.97-6.97a.75.75 0 111.06 1.06L13.06 12l6.97 6.97a.75.75 0 11-1.06 1.06L12 13.06l-6.97 6.97a.75.75 0 01-1.06-1.06L10.94 12 3.97 5.03a.75.75 0 010-1.06z"
                fill-rule="evenodd"
              />
            </svg>
             
            At least 1 uppercase letter
          </div>
          <div
            class="validator no-match"
          >
            <svg
              aria-hidden="true"
              class="icon x"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M3.97 3.97a.75.75 0 011.06 0L12 10.94l6.97-6.97a.75.75 0 111.06 1.06L13.06 12l6.97 6.97a.75.75 0 11-1.06 1.06L12 13.06l-6.97 6.97a.75.75 0 01-1.06-1.06L10.94 12 3.97 5.03a.75.75 0 010-1.06z"
                fill-rule="evenodd"
              />
            </svg>
             
            At least 1 number
          </div>
          <div
            class="validator no-match"
          >
            <svg
              aria-hidden="true"
              class="icon x"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M3.97 3.97a.75.75 0 011.06 0L12 10.94l6.97-6.97a.75.75 0 111.06 1.06L13.06 12l6.97 6.97a.75.75 0 11-1.06 1.06L12 13.06l-6.97 6.97a.75.75 0 01-1.06-1.06L10.94 12 3.97 5.03a.75.75 0 010-1.06z"
                fill-rule="evenodd"
              />
            </svg>
             
            At least 1 special character
          </div>
        </div>
      </div>
      <div
        class="input-and-error-container"
        data-testid="text-input"
      >
        <div
          class="input-and-label-container"
        >
          <div
            class="input-container"
          >
            <input
              class="input"
              id="confirmPassword"
              name="confirmPassword"
              type="password"
              value=""
            />
            <svg
              aria-hidden="false"
              aria-label="Show Confirm Password"
              class="eye"
              fill="currentColor"
              tabindex="0"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 12.5a2.5 2.5 0 100-5 2.5 2.5 0 000 5z"
              />
              <path
                clip-rule="evenodd"
                d="M.664 10.59a1.651 1.651 0 010-1.186A10.004 10.004 0 0110 3c4.257 0 7.893 2.66 9.336 6.41.147.381.146.804 0 1.186A10.004 10.004 0 0110 17c-4.257 0-7.893-2.66-9.336-6.41zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
          <label
            class="label"
            for="confirmPassword"
          >
            Confirm Password
          </label>
        </div>
      </div>
      <div
        style="margin-top: 1rem;"
      >
        <button
          class="button positive"
          style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
          type="submit"
        >
          Create new password
        </button>
      </div>
    </form>
  </div>
</main>
`;
